frontend/device/gesturedetector.lua

```diff
@@ -461,30 +501,20 @@ function GestureDetector:handleNonTap(tev)
         -- switched from other state, probably from initialState
         -- we return nil in this case
         self.states[slot] = self.tapState
-        if self._has_real_clock_time_ev_time == nil then
-            if tev.timev.sec < TimeVal:now().sec - 600 then
-                -- ev.timev is probably the uptime since device boot
-                -- (which might pause on suspend) that we can't use
-                -- with setTimeout(): we'll use TimeVal:now()
-                self._has_real_clock_time_ev_time = false
-                logger.info("event times are not real clock time: some adjustments will be made")
-            else
-                -- assume they are real clock time
-                self._has_real_clock_time_ev_time = true
-                logger.info("event times are real clock time: no adjustment needed")
-            end
-        end
         logger.dbg("set up hold timer")
-        local ref_time = self._has_real_clock_time_ev_time and tev.timev or TimeVal:now()
-        local deadline = ref_time + TimeVal:new{
-            sec = 0, usec = ges_hold_interval
-        }
-        -- Be sure the following setTimeout only react to this tapState
-        local hold_timer_id = tev.timev
-        self.hold_timer_id[slot] = hold_timer_id
-        self.input:setTimeout(function()
-            if self.states[slot] == self.tapState and self.hold_timer_id[slot] == hold_timer_id then
-                -- timer set in tapState, so we switch to hold
+        -- Ditto, start counting from *now*, no matter what the input event's timestamp says...
+        local deadline = TimeVal:now() + ges_hold_interval
+        print("GestureDetector:handleNonTap: Updating pending_hold_timer for slot", slot)
+        -- Invalidate previous hold timers on that slot so that the following setTimeout will only react to *this* tapState.
+        self.input:clearTimeout(slot, "hold")
+        self.pending_hold_timer[slot] = true
+        self.input:setTimeout(slot, "hold", function()
+            print("GestureDetector: Hold finalization for slot", slot)
+            -- If the pending_hold_timer we set on our first switch to tapState on this slot (e.g., first finger down event),
+            -- back when the timer was setup, is still relevant (e.g., the slot wasn't run through clearState by a finger up gesture),
+            -- then check that we're still in a stationary finger down state (e.g., tapState).
+            if self.pending_hold_timer[slot] and self.states[slot] == self.tapState then
```

Comment:
You're sure this is ok ? Just having self.pending_hold_timer[slot] = true ?
You clear it, then reset-it to true: a previous setTimeout'ed will see it to true and will proceed.

The previous code was using a unique ID (tev.timev) to have a previous setTimeout just know it's no more its job.

Reply:
> You clear it, then reset-it to true: a previous setTimeout'ed will see it to true and will proceed.

But it won't, because the clearTimeout means that there's no longer a previous setTimeout ;).

Reply:
Oh, I see, you tweaked setTimeout so there are no longer "anynoymous function", but they get a name/slot internally.
And well, I guess I thought this was managed by UIManager like what we scheduleIn() - so, ok.

Reply:
Yeah, that allowed to clean them up early based on what they were for ;).

(Granted, the only time I actually could end up with a stack of two or three pending hold timers that would never ever be true was mashing on the virtual keyboard like a crazy person ^^).

Reply:
Found the PR I made with these timer ids, where I explained the issue I had and how to reproduce it: #6186

Reply:
Ha, thanks ;).

Yep, still fixed (I did manage to inadvertently set the screenshot as the screensaver, which should prove that it behaves ;p).

Reply:
Can you reproduce it if you remove your timeout clearing ?

Reply:
Most likely, yes.